org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/workspace/BaseConfigurator.java

```diff
@@ -0,0 +1,128 @@
+        progress.setWorkRemaining(4);
+
+        PersistentModelBuilder persistentModel = new PersistentModelBuilder(CorePlugin.modelPersistence().loadModel(project));
+        EclipseProject model = lookupEclipseModel(project);
+        progress.worked(1);
+
+        BuildScriptLocationUpdater.update(model, persistentModel, progress.newChild(1));
+
+        LinkedResourcesUpdater.update(project, ImmutableList.copyOf(model.getLinkedResources()), persistentModel, progress.newChild(1));
+        GradleFolderUpdater.update(project, model, persistentModel, progress.newChild(1));
+        ProjectNatureUpdater.update(project,  ImmutableList.copyOf(model.getProjectNatures()), persistentModel, progress.newChild(1));
+        BuildCommandUpdater.update(project, ImmutableList.copyOf(model.getBuildCommands()), persistentModel, progress.newChild(1));
+
+        if (isJavaProject(model)) {
+            synchronizeJavaProject(model, project, persistentModel, progress);
+        } else {
+            persistentModel.classpath(ImmutableList.<IClasspathEntry>of());
+        }
+
+        CorePlugin.modelPersistence().saveModel(persistentModel.build());
+    }
+
+    private void synchronizeJavaProject(final EclipseProject model, final IProject project, final PersistentModelBuilder persistentModel, SubMonitor progress) throws CoreException {
+        JavaCore.run(new IWorkspaceRunnable() {
+            @Override
+            public void run(IProgressMonitor monitor) throws CoreException {
+                SubMonitor progress = SubMonitor.convert(monitor);
+                synchronizeJavaProjectInTransaction(model, project, persistentModel, progress);
+            }
+        }, progress.newChild(1));
+    }
+
+    private void synchronizeJavaProjectInTransaction(final EclipseProject model, final IProject project, PersistentModelBuilder persistentModel, SubMonitor progress) throws JavaModelException, CoreException {
+        progress.setWorkRemaining(8);
+        //old Gradle versions did not expose natures, so we need to add the Java nature explicitly
+        CorePlugin.workspaceOperations().addNature(project, JavaCore.NATURE_ID, progress.newChild(1));
+        IJavaProject javaProject = JavaCore.create(project);
+        OutputLocationUpdater.update(javaProject, model.getOutputLocation(), progress.newChild(1));
+        SourceFolderUpdater.update(javaProject, ImmutableList.copyOf(model.getSourceDirectories()), progress.newChild(1));
+        LibraryFilter.update(javaProject, model, progress.newChild(1));
+        ClasspathContainerUpdater.update(javaProject, model, progress.newChild(1));
+        JavaSourceSettingsUpdater.update(javaProject, model, progress.newChild(1));
+        GradleClasspathContainerUpdater.updateFromModel(javaProject, model, this.locationToProject.values(), persistentModel, progress.newChild(1));
+        WtpClasspathUpdater.update(javaProject, model, this.internalGradleBuild, progress.newChild(1));
```

Comment:
⭕️ Should probably be in its own wtp configurator, just to illustrate the point of configurators.

Reply:
👍